ipie/addons/ephqmc/trial_wavefunction/variational/coherent_state_variational.py

```diff
@@ -0,0 +1,173 @@
+import numpy as np
+from scipy.optimize import basinhopping
+
+from ipie.legacy.trial_wavefunction.harmonic_oscillator import HarmonicOscillator
+from ipie.systems.generic import Generic
+from ipie.addons.eph.hamiltonians.holstein import HolsteinModel
+
+import jax
+
+def local_energy(
+        X: np.ndarray,
+        G: np.ndarray,
+        m: float,
+        w: float,
+        g: float,
+        nsites: int,
+        Lap: np.ndarray,
+        T: np.ndarray,
+        nup: int,
+        ndown: int
+    ) -> np.ndarray:
+
+    kinetic_contrib = jax.numpy.einsum('ij->', T[0] * G[0])
+    if ndown > 0:
+        kinetic_contrib += jax.numpy.einsum('ij->', T[1] * G[1])
+
+    rho = G[0].diagonal() + G[1].diagonal()
+    el_ph_contrib = -g * np.sqrt(2 * m * w) * np.sum(rho * X)
+
+    phonon_contrib = m * w**2 * np.sum(X * X) / 2
+    phonon_contrib += -0.5 * np.sum(Lap) / m - 0.5 * w * nsites
+
+    local_energy = kinetic_contrib + el_ph_contrib + phonon_contrib
+    return local_energy
+
+def gab(A: np.ndarray, B: np.ndarray):
```

Comment:
this is also defined in the towazaya. Maybe consolidate repeated code in variational/estimators.py which has all the jax specific code.